File: backend/app/schemas/interview.py
"""面试准备模块（M5，需求 3.5）Pydantic 模型。

- 面试题生成：技术题 / 项目题 / 行为题三类
- 模拟面试：会话 + 多轮问答
- 答案评估：逻辑性 / 完整性 / 专业性三维度评分 + 优点/不足/改进建议
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field

from app.schemas.common import ORMBase


class InterviewCategory(str, Enum):
    """面试题类型（需求 3.5：技术题、项目题、行为题）"""

    TECHNICAL = "technical"  # 技术题
    PROJECT = "project"  # 项目题
    BEHAVIORAL = "behavioral"  # 行为题


class InterviewCreate(BaseModel):
    """创建模拟面试会话请求"""

    job_id: int = Field(..., description="目标岗位ID（已解析的 job_analysis）")
    resume_id: Optional[int] = Field(None, description="简历ID，缺省取用户最新简历")
    question_count: int = Field(6, ge=3, le=9, description="题目数量（3-9，三类尽量均匀）")


class QuestionGenerateRequest(BaseModel):
    """仅生成面试题（不开面试会话）请求"""

    job_id: int = Field(..., description="目标岗位ID")
    resume_id: Optional[int] = Field(None, description="简历ID，缺省取用户最新简历")
    question_count: int = Field(6, ge=3, le=9)


class GeneratedQuestion(BaseModel):
    """LLM 生成的单道面试题（结构化输出校验，需求 8.3）"""

    category: InterviewCategory
    question: str = Field(..., min_length=2)


class QAOut(ORMBase):
    """单轮问答记录输出"""

    id: int
    session_id: int
    category: str
    question: str
    answer: Optional[str] = None
    score: Optional[int] = Field(None, description="本轮综合得分 0-100")
    feedback: Optional[str] = None
    feedback_json: Optional[dict] = Field(
        None, description="三维度评分与优点/不足/建议明细"
    )


class SessionOut(BaseModel):
    """面试会话详情输出（含全部问答）"""

    id: int
    user_id: int
    job_id: Optional[int] = None
    resume_id: Optional[int] = None
    position_title: Optional[str] = None
    status: str
    summary: Optional[str] = Field(None, description="面试结束后的总评报告")
    created_at: datetime
    finished_at: Optional[datetime] = None
    qas: List[QAOut] = Field(default_factory=list)


class SessionListItem(BaseModel):
    """面试会话列表项"""

    id: int
    job_id: Optional[int] = None
    position_title: Optional[str] = None
    status: str
    created_at: datetime
    finished_at: Optional[datetime] = None
    qa_count: int
    answered_count: int
    average_score: Optional[float] = None


class AnswerSubmit(BaseModel):
    """提交一轮回答"""

    qa_id: int = Field(..., description="本轮问答记录ID")
    answer: str = Field(..., min_length=1, description="用户回答文本")


class AnswerResult(BaseModel):
    """提交回答后的返回：本轮评估 + 下一道待答题"""

    qa: QAOut
    next_qa: Optional[QAOut] = None
    is_finished: bool = Field(..., description="是否已无待答题（可结束面试生成总评）")


class QuestionBankResponse(BaseModel):
    """面试题生成结果"""

    job_id: int
    position_title: Optional[str] = None
    questions: List[GeneratedQuestion] = Field(default_factory=list)
    source: str = Field("llm", description="题目来源：llm / fallback")

File: frontend/src/api/interview.ts
import api from './index'

// 面试题类型（M5：技术题 / 项目题 / 行为题）
export type InterviewCategory = 'technical' | 'project' | 'behavioral'

export interface GeneratedQuestion {
  category: InterviewCategory
  question: string
}

// 单轮回答的多维度评估
export interface FeedbackJSON {
  scores: {
    logic: number
    completeness: number
    professionalism: number
  }
  overall: number
  strengths: string
  weaknesses: string
  suggestions: string
  source: 'llm' | 'fallback'
}

export interface InterviewQA {
  id: number
  session_id: number
  category: string
  question: string
  answer: string | null
  score: number | null
  feedback: string | null
  feedback_json: FeedbackJSON | null
}

export interface InterviewSession {
  id: number
  user_id: number
  job_id: number | null
  resume_id: number | null
  position_title?: string | null
  status: 'active' | 'finished' | string
  summary?: string | null
  created_at: string
  finished_at?: string | null
  qas: InterviewQA[]
}

export interface SessionListItem {
  id: number
  job_id: number | null
  position_title?: string | null
  status: string
  created_at: string
  finished_at?: string | null
  qa_count: number
  answered_count: number
  average_score: number | null
}

export interface AnswerResult {
  qa: InterviewQA
  next_qa: InterviewQA | null
  is_finished: boolean
}

/** 创建模拟面试会话（同时由 LLM 生成题库，耗时较长） */
export function createInterviewSession(data: {
  job_id: number
  resume_id?: number
  question_count?: number
}): Promise<InterviewSession> {
  return api
    .post<InterviewSession>('/interview/session', data, { timeout: 180000 })
    .then(r => r.data)
}

/** 获取面试会话历史列表 */
export function listInterviewSessions(): Promise<SessionListItem[]> {
  return api.get<SessionListItem[]>('/interview/sessions').then(r => r.data)
}

/** 获取面试会话详情（含全部问答与评估） */
export function getInterviewSession(id: number): Promise<InterviewSession> {
  return api.get<InterviewSession>(`/interview/session/${id}`).then(r => r.data)
}

/** 提交一轮回答，返回三维度评估与下一道题 */
export function submitAnswer(
  sessionId: number,
  qaId: number,
  answer: string
): Promise<AnswerResult> {
  return api
    .post<AnswerResult>(
      `/interview/session/${sessionId}/answer`,
      { qa_id: qaId, answer },
      { timeout: 120000 }
    )
    .then(r => r.data)
}

/** 结束面试并生成总评报告 */
export function finishInterview(sessionId: number): Promise<InterviewSession> {
  return api
    .post<InterviewSession>(`/interview/session/${sessionId}/finish`, {}, { timeout: 180000 })
    .then(r => r.data)
}

/** 仅生成面试题（不开面试会话） */
export function generateInterviewQuestions(data: {
  job_id: number
  resume_id?: number
  question_count?: number
}): Promise<{
  job_id: number
  position_title?: string | null
  questions: GeneratedQuestion[]
  source: string
}> {
  return api
    .post('/interview/questions', data, { timeout: 180000 })
    .then(r => r.data)
}

// 题型展示配置
export const categoryMeta: Record<string, { label: string; color: string }> = {
  technical: { label: '技术题', color: '#667eea' },
  project: { label: '项目题', color: '#764ba2' },
  behavioral: { label: '行为题', color: '#f093fb' },
}
